library/x509_crt.c

```diff
@@ -657,8 +659,11 @@ static int x509_get_subject_alt_name( unsigned char **p,
                     MBEDTLS_ERR_ASN1_UNEXPECTED_TAG );
         }
 
-        /* Skip everything but DNS name */
-        if( tag != ( MBEDTLS_ASN1_CONTEXT_SPECIFIC | 2 ) )
+        tag_mask = tag & ( MBEDTLS_ASN1_TAG_CLASS_MASK | MBEDTLS_ASN1_TAG_VALUE_MASK );
+
+        /* Skip everything but DNS name and Other Name */
+        if( tag_mask != ( MBEDTLS_ASN1_CONTEXT_SPECIFIC | 2 ) &&
```

Comment:
As we're not parsing the name components here, why don't we just return the linked list of all of them, and leave it to the user to ignore whatever he's not interested in?

Reply:
but we are parsing the name compoenent here, at least the DNS name

Reply:
@RonEld We don't parse the extension in the sense that we just return the raw buffer, but don't look interpret its contents any further - but OTOH, there's not much to parse for `dNSName` entries in the first place, so my argument is probably weak.

Nevertheless, I would like to maintain my suggestion that the `SubjectAlternativeName` parsing should just return the raw list of components, comprising ASN.1 tag + raw data, leaving it to the user to do further inspection. That doesn't mean we can't provide parsing functions and structures for specific name types, but I don't want them to be used by default to avoid the automatic increase of code-size and memory usage, even for those users that don't make use of them.

This principle is already partly illustrated in #2478: The PR removes any dynamically allocated linked lists from `mbedtls_x509_crt`, confining to raw data buffers, but does provide convenience functions for the user to build the 'old' linked list structures from the raw data, if they need it.

In the same vein, I would like to suggest that this PR -- as long as it's not based on #2478 -- just adds `HardwareModuleName` to the `subject_alt_names` linked list, and potentially but not necessarily provides convenience functions for further parsing of these.

Reply:
I totally agree, but this PR keeps the same existing architecture.
If we change this, it will divert from the existing architecture towards #2478.
In the long run, this is what's going to happen, but for the sake of existing design, we shouldn't change, IMHO

Reply:
Sorry @RonEld, I am not sure where is our divergence? My proposal is also aligned with the current architecture, isn't it?

Reply:
As I understand your suggestion, you are suggesting to put the sequence raw data in a buffer, without reading and parsing it.
Current behavior is we read the contents name. Am I missing something?

Reply:
@RonEld We're currently providing a linked list of the ASN.1 `Sequence` entries, but don't further parse these entries; I am suggesting that we keep it that way, even if the name components are structured (such as for `HardwareModuleName`).

Reply:
Actually, this is what I did for `otherName`, I just didn't enhance it further for other types, as I think that we should also supply the parsing function to support other types.
I prefer to keep it maintained in what we support parsing

Reply:
In addition, current behavior is not to create a new `mbedtls_asn1_sequence` for every supported item. If we just call `mbedtls_asn1_get_sequence_of()`, it will result with additional memory being allocated